File: javascript/pythagorean-triplet/pythagorean-triplet.js
export class Triplet {
    constructor(a, b, c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }

    sum() {
        return this.a + this.b + this.c;
    }

    product() {
        return this.a * this.b * this.c;
    }

    isPythagorean() {
        return this.a < this.b && this.b < this.c && (this.a**2 + this.b**2) === this.c**2;
    }

    static where(criteria) {
        const triplets = [];
        const minFactor = criteria.minFactor || 1;
        const maxFactor = criteria.maxFactor || 1000;
        const matchesSum = (triplet) => criteria.sum ? triplet.sum() === criteria.sum : true;
        for (let c = minFactor + 2; c <= maxFactor; c++) {
            for (let b = minFactor + 1; b < c; b++) {
                for (let a = minFactor; a < b; a++) {
                    const triplet = new Triplet(a, b, c);
                    if (triplet.isPythagorean() && matchesSum(triplet)) {
                        triplets.push(triplet);
                    }
                }
            }
        }
        return triplets;
    }

    toString() {
        return this.product().toString().padStart(16, '0');
    }
}

File: ruby/pythagorean-triplet/_iterations/3/pythagorean_triplet.rb
class Triplet
  attr_reader :factors

  def initialize(*factors)
    @factors = factors
  end

  def sum
    factors.sum
  end

  def product
    factors.reduce(1, &:*)
  end

  def pythagorean?
    factors[0, factors.size - 1].map { |n| n ** 2 }.sum == factors.last ** 2 
  end

  def self.where(min_factor: 1, max_factor:, sum: nil)
    (min_factor..(max_factor - 2)).flat_map do |f0|
      ((f0 + 1)..(max_factor - 1)).flat_map do |f1|
        if sum.nil?
          ((f1 + 1)..max_factor).map { |f2| new(f0, f1, f2) }
        else
          f2 = sum - f0 - f1
          f2 <= max_factor ? [new(f0, f1, f2)] : []
        end
      end
    end.filter(&:pythagorean?)
  end
end
